File: front/central/src/services/modules/inventory/app/use-cases.test.ts
import { describe, it, expect, vi, beforeEach } from 'vitest';
import { InventoryUseCases } from './use-cases';
import { IInventoryRepository } from '../domain/ports';
import {
    InventoryLevel,
    StockMovement,
    MovementType,
    InventoryListResponse,
    MovementListResponse,
    MovementTypeListResponse,
    AdjustStockDTO,
    TransferStockDTO,
    GetInventoryParams,
    GetMovementsParams,
} from '../domain/types';

// -----------------------------------------------------------------
// Helpers: datos de prueba reutilizables
// -----------------------------------------------------------------

const makeInventoryLevel = (overrides: Partial<InventoryLevel> = {}): InventoryLevel => ({
    id: 1,
    product_id: 'prod-001',
    warehouse_id: 1,
    location_id: null,
    business_id: 1,
    quantity: 100,
    reserved_qty: 10,
    available_qty: 90,
    min_stock: 5,
    max_stock: 500,
    reorder_point: 20,
    product_name: 'Producto de prueba',
    product_sku: 'SKU-001',
    warehouse_name: 'Bodega principal',
    warehouse_code: 'WH-01',
    created_at: '2026-01-01T00:00:00Z',
    updated_at: '2026-01-01T00:00:00Z',
    ...overrides,
});

const makeStockMovement = (overrides: Partial<StockMovement> = {}): StockMovement => ({
    id: 1,
    product_id: 'prod-001',
    warehouse_id: 1,
    location_id: null,
    business_id: 1,
    movement_type_id: 1,
    movement_type_code: 'adjustment',
    movement_type_name: 'Ajuste',
    reason: 'Ajuste de inventario',
    quantity: 10,
    previous_qty: 90,
    new_qty: 100,
    reference_type: null,
    reference_id: null,
    integration_id: null,
    notes: 'Ajuste manual',
    created_by_id: 1,
    product_name: 'Producto de prueba',
    product_sku: 'SKU-001',
    warehouse_name: 'Bodega principal',
    created_at: '2026-01-01T00:00:00Z',
    ...overrides,
});

const makeMovementType = (overrides: Partial<MovementType> = {}): MovementType => ({
    id: 1,
    code: 'adjustment',
    name: 'Ajuste',
    description: 'Ajuste de inventario',
    is_active: true,
    direction: 'in',
    created_at: '2026-01-01T00:00:00Z',
    updated_at: '2026-01-01T00:00:00Z',
    ...overrides,
});

const makeInventoryListResponse = (overrides: Partial<InventoryListResponse> = {}): InventoryListResponse => ({
    data: [makeInventoryLevel()],
    total: 1,
    page: 1,
    page_size: 10,
    total_pages: 1,
    ...overrides,
});

const makeMovementListResponse = (overrides: Partial<MovementListResponse> = {}): MovementListResponse => ({
    data: [makeStockMovement()],
    total: 1,
    page: 1,
    page_size: 10,
    total_pages: 1,
    ...overrides,
});

const makeMovementTypeListResponse = (overrides: Partial<MovementTypeListResponse> = {}): MovementTypeListResponse => ({
    data: [makeMovementType()],
    total: 1,
    page: 1,
    page_size: 10,
    total_pages: 1,
    ...overrides,
});

// -----------------------------------------------------------------
// Mock del repositorio
// -----------------------------------------------------------------

function createMockRepository(): IInventoryRepository {
    return {
        getProductInventory: vi.fn(),
        getWarehouseInventory: vi.fn(),
        adjustStock: vi.fn(),
        transferStock: vi.fn(),
        getMovements: vi.fn(),
        getMovementTypes: vi.fn(),
    } as unknown as IInventoryRepository;
}

// -----------------------------------------------------------------
// Suite principal
// -----------------------------------------------------------------

describe('InventoryUseCases', () => {
    let repo: ReturnType<typeof createMockRepository>;
    let useCases: InventoryUseCases;

    beforeEach(() => {
        repo = createMockRepository();
        useCases = new InventoryUseCases(repo as unknown as IInventoryRepository);
    });

    // ---------------------------------------------------------------
    // getProductInventory
    // ---------------------------------------------------------------
    describe('getProductInventory', () => {
        it('debería retornar el inventario de un producto', async () => {
            const inventoryLevels = [makeInventoryLevel()];
            vi.mocked(repo.getProductInventory).mockResolvedValue(inventoryLevels);

            const result = await useCases.getProductInventory('prod-001');

            expect(result).toEqual(inventoryLevels);
            expect(repo.getProductInventory).toHaveBeenCalledOnce();
            expect(repo.getProductInventory).toHaveBeenCalledWith('prod-001', undefined);
        });

        it('debería pasar el businessId cuando se proporciona', async () => {
            const inventoryLevels = [makeInventoryLevel()];
            vi.mocked(repo.getProductInventory).mockResolvedValue(inventoryLevels);

            await useCases.getProductInventory('prod-001', 5);

            expect(repo.getProductInventory).toHaveBeenCalledWith('prod-001', 5);
        });

        it('debería propagar el error cuando el repositorio falla', async () => {
            const expectedError = new Error('Producto no encontrado');
            vi.mocked(repo.getProductInventory).mockRejectedValue(expectedError);

            await expect(useCases.getProductInventory('prod-999')).rejects.toThrow('Producto no encontrado');
        });
    });

    // ---------------------------------------------------------------
    // getWarehouseInventory
    // ---------------------------------------------------------------
    describe('getWarehouseInventory', () => {
        it('debería retornar el inventario de una bodega con parámetros', async () => {
            const response = makeInventoryListResponse();
            vi.mocked(repo.getWarehouseInventory).mockResolvedValue(response);

            const params: GetInventoryParams = { page: 1, page_size: 10 };
            const result = await useCases.getWarehouseInventory(1, params);

            expect(result).toEqual(response);
            expect(repo.getWarehouseInventory).toHaveBeenCalledWith(1, params);
        });

        it('debería llamar al repositorio sin parámetros cuando no se pasan', async () => {
            const response = makeInventoryListResponse();
            vi.mocked(repo.getWarehouseInventory).mockResolvedValue(response);

            await useCases.getWarehouseInventory(1);

            expect(repo.getWarehouseInventory).toHaveBeenCalledWith(1, undefined);
        });

        it('debería propagar el error cuando el repositorio falla', async () => {
            const expectedError = new Error('Bodega no encontrada');
            vi.mocked(repo.getWarehouseInventory).mockRejectedValue(expectedError);

            await expect(useCases.getWarehouseInventory(99)).rejects.toThrow('Bodega no encontrada');
        });
    });

    // ---------------------------------------------------------------
    // adjustStock
    // ---------------------------------------------------------------
    describe('adjustStock', () => {
        const adjustDto: AdjustStockDTO = {
            product_id: 'prod-001',
            warehouse_id: 1,
            quantity: 10,
            reason: 'Ajuste manual',
            notes: 'Corrección de conteo',
        };

        it('debería ajustar el stock y retornar el movimiento creado', async () => {
            const movement = makeStockMovement();
            vi.mocked(repo.adjustStock).mockResolvedValue(movement);

            const result = await useCases.adjustStock(adjustDto);

            expect(result).toEqual(movement);
            expect(repo.adjustStock).toHaveBeenCalledOnce();
            expect(repo.adjustStock).toHaveBeenCalledWith(adjustDto, undefined);
        });

        it('debería pasar el businessId cuando se proporciona', async () => {
            const movement = makeStockMovement();
            vi.mocked(repo.adjustStock).mockResolvedValue(movement);

            await useCases.adjustStock(adjustDto, 5);

            expect(repo.adjustStock).toHaveBeenCalledWith(adjustDto, 5);
        });

        it('debería propagar el error cuando el ajuste falla', async () => {
            const expectedError = new Error('Stock insuficiente');
            vi.mocked(repo.adjustStock).mockRejectedValue(expectedError);

            await expect(useCases.adjustStock(adjustDto)).rejects.toThrow('Stock insuficiente');
        });
    });

    // ---------------------------------------------------------------
    // transferStock
    // ---------------------------------------------------------------
    describe('transferStock', () => {
        const transferDto: TransferStockDTO = {
            product_id: 'prod-001',
            from_warehouse_id: 1,
            to_warehouse_id: 2,
            quantity: 5,
            reason: 'Reabastecimiento',
        };

        it('debería transferir stock y retornar el mensaje de confirmación', async () => {
            const response = { message: 'Transferencia exitosa' };
            vi.mocked(repo.transferStock).mockResolvedValue(response);

            const result = await useCases.transferStock(transferDto);

            expect(result).toEqual(response);
            expect(repo.transferStock).toHaveBeenCalledOnce();
            expect(repo.transferStock).toHaveBeenCalledWith(transferDto, undefined);
        });

        it('debería pasar el businessId cuando se proporciona', async () => {
            const response = { message: 'Transferencia exitosa' };
            vi.mocked(repo.transferStock).mockResolvedValue(response);

            await useCases.transferStock(transferDto, 3);

            expect(repo.transferStock).toHaveBeenCalledWith(transferDto, 3);
        });

        it('debería propagar el error cuando la transferencia falla', async () => {
            const expectedError = new Error('Cantidad excede el stock disponible');
            vi.mocked(repo.transferStock).mockRejectedValue(expectedError);

            await expect(useCases.transferStock(transferDto)).rejects.toThrow('Cantidad excede el stock disponible');
        });
    });

    // ---------------------------------------------------------------
    // getMovements
    // ---------------------------------------------------------------
    describe('getMovements', () => {
        it('debería retornar la lista paginada de movimientos cuando se pasan parámetros', async () => {
            const response = makeMovementListResponse();
            vi.mocked(repo.getMovements).mockResolvedValue(response);

            const params: GetMovementsParams = { page: 1, page_size: 10, type: 'adjustment' };
            const result = await useCases.getMovements(params);

            expect(result).toEqual(response);
            expect(repo.getMovements).toHaveBeenCalledOnce();
            expect(repo.getMovements).toHaveBeenCalledWith(params);
        });

        it('debería llamar al repositorio sin parámetros cuando no se pasan', async () => {
            const response = makeMovementListResponse();
            vi.mocked(repo.getMovements).mockResolvedValue(response);

            await useCases.getMovements();

            expect(repo.getMovements).toHaveBeenCalledWith(undefined);
        });

        it('debería propagar el error cuando el repositorio falla', async () => {
            const expectedError = new Error('Error de base de datos');
            vi.mocked(repo.getMovements).mockRejectedValue(expectedError);

            await expect(useCases.getMovements()).rejects.toThrow('Error de base de datos');
        });
    });

    // ---------------------------------------------------------------
    // getMovementTypes
    // ---------------------------------------------------------------
    describe('getMovementTypes', () => {
        it('debería retornar la lista de tipos de movimiento con parámetros', async () => {
            const response = makeMovementTypeListResponse();
            vi.mocked(repo.getMovementTypes).mockResolvedValue(response);

            const params = { page: 1, page_size: 10, active_only: true };
            const result = await useCases.getMovementTypes(params);

            expect(result).toEqual(response);
            expect(repo.getMovementTypes).toHaveBeenCalledOnce();
            expect(repo.getMovementTypes).toHaveBeenCalledWith(params);
        });

        it('debería llamar al repositorio sin parámetros cuando no se pasan', async () => {
            const response = makeMovementTypeListResponse();
            vi.mocked(repo.getMovementTypes).mockResolvedValue(response);

            await useCases.getMovementTypes();

            expect(repo.getMovementTypes).toHaveBeenCalledWith(undefined);
        });

        it('debería propagar el error cuando el repositorio falla', async () => {
            const expectedError = new Error('Servicio no disponible');
            vi.mocked(repo.getMovementTypes).mockRejectedValue(expectedError);

            await expect(useCases.getMovementTypes()).rejects.toThrow('Servicio no disponible');
        });
    });
});

File: mobile/mobile_central/test/services/modules/inventory/app/use_cases_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:mobile_central/services/modules/inventory/app/use_cases.dart';
import 'package:mobile_central/services/modules/inventory/domain/entities.dart';
import 'package:mobile_central/services/modules/inventory/domain/ports.dart';
import 'package:mobile_central/shared/types/paginated_response.dart';

// ---------------------------------------------------------------------------
// Manual mock for IInventoryRepository
// ---------------------------------------------------------------------------
class MockInventoryRepository implements IInventoryRepository {
  final List<String> calls = [];

  List<InventoryLevel>? getProductInventoryResult;
  PaginatedResponse<InventoryLevel>? getWarehouseInventoryResult;
  StockMovement? adjustStockResult;
  Map<String, dynamic>? transferStockResult;
  PaginatedResponse<StockMovement>? getMovementsResult;
  PaginatedResponse<MovementType>? getMovementTypesResult;

  Exception? errorToThrow;

  // Captured args
  String? lastProductId;
  int? lastWarehouseId;
  int? lastBusinessId;
  GetInventoryParams? lastInventoryParams;
  AdjustStockDTO? lastAdjustDTO;
  TransferStockDTO? lastTransferDTO;
  GetMovementsParams? lastMovementsParams;
  GetMovementTypesParams? lastMovementTypesParams;

  void _trackCall(String name) {
    calls.add(name);
    if (errorToThrow != null) throw errorToThrow!;
  }

  @override
  Future<List<InventoryLevel>> getProductInventory(String productId,
      {int? businessId}) async {
    lastProductId = productId;
    lastBusinessId = businessId;
    _trackCall('getProductInventory');
    return getProductInventoryResult!;
  }

  @override
  Future<PaginatedResponse<InventoryLevel>> getWarehouseInventory(
      int warehouseId, GetInventoryParams? params) async {
    lastWarehouseId = warehouseId;
    lastInventoryParams = params;
    _trackCall('getWarehouseInventory');
    return getWarehouseInventoryResult!;
  }

  @override
  Future<StockMovement> adjustStock(AdjustStockDTO data,
      {int? businessId}) async {
    lastAdjustDTO = data;
    lastBusinessId = businessId;
    _trackCall('adjustStock');
    return adjustStockResult!;
  }

  @override
  Future<Map<String, dynamic>> transferStock(TransferStockDTO data,
      {int? businessId}) async {
    lastTransferDTO = data;
    lastBusinessId = businessId;
    _trackCall('transferStock');
    return transferStockResult!;
  }

  @override
  Future<PaginatedResponse<StockMovement>> getMovements(
      GetMovementsParams? params) async {
    lastMovementsParams = params;
    _trackCall('getMovements');
    return getMovementsResult!;
  }

  @override
  Future<PaginatedResponse<MovementType>> getMovementTypes(
      GetMovementTypesParams? params) async {
    lastMovementTypesParams = params;
    _trackCall('getMovementTypes');
    return getMovementTypesResult!;
  }
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------
Pagination _makePagination() {
  return Pagination(
    currentPage: 1,
    perPage: 10,
    total: 1,
    lastPage: 1,
    hasNext: false,
    hasPrev: false,
  );
}

InventoryLevel _makeInventoryLevel({int id = 1}) {
  return InventoryLevel(
    id: id,
    productId: 'prod-$id',
    warehouseId: 1,
    businessId: 1,
    quantity: 100,
    reservedQty: 10,
    availableQty: 90,
    createdAt: '',
    updatedAt: '',
  );
}

StockMovement _makeMovement({int id = 1}) {
  return StockMovement(
    id: id,
    productId: 'prod-$id',
    warehouseId: 1,
    businessId: 1,
    movementTypeId: 1,
    movementTypeCode: 'ADJ',
    movementTypeName: 'Adjustment',
    reason: 'test',
    quantity: 10,
    previousQty: 90,
    newQty: 100,
    notes: '',
    createdAt: '',
  );
}

MovementType _makeMovementType({int id = 1}) {
  return MovementType(
    id: id,
    code: 'ADJ',
    name: 'Adjustment',
    description: 'Stock adjustment',
    isActive: true,
    direction: 'in',
    createdAt: '',
    updatedAt: '',
  );
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------
void main() {
  late MockInventoryRepository mockRepo;
  late InventoryUseCases useCases;

  setUp(() {
    mockRepo = MockInventoryRepository();
    useCases = InventoryUseCases(mockRepo);
  });

  group('getProductInventory', () {
    test('delegates to repository and returns result', () async {
      mockRepo.getProductInventoryResult = [_makeInventoryLevel(id: 1)];

      final result =
          await useCases.getProductInventory('prod-1', businessId: 5);

      expect(mockRepo.calls, ['getProductInventory']);
      expect(mockRepo.lastProductId, 'prod-1');
      expect(mockRepo.lastBusinessId, 5);
      expect(result.length, 1);
      expect(result.first.id, 1);
    });

    test('propagates error from repository', () {
      mockRepo.errorToThrow = Exception('network error');

      expect(
        () => useCases.getProductInventory('prod-1'),
        throwsA(isA<Exception>()),
      );
    });
  });

  group('getWarehouseInventory', () {
    test('delegates to repository and returns result', () async {
      mockRepo.getWarehouseInventoryResult = PaginatedResponse<InventoryLevel>(
        data: [_makeInventoryLevel()],
        pagination: _makePagination(),
      );

      final params = GetInventoryParams(page: 1, pageSize: 10);
      final result = await useCases.getWarehouseInventory(5, params);

      expect(mockRepo.calls, ['getWarehouseInventory']);
      expect(mockRepo.lastWarehouseId, 5);
      expect(mockRepo.lastInventoryParams, params);
      expect(result.data.length, 1);
    });

    test('delegates with null params', () async {
      mockRepo.getWarehouseInventoryResult = PaginatedResponse<InventoryLevel>(
        data: [],
        pagination: _makePagination(),
      );

      await useCases.getWarehouseInventory(1, null);

      expect(mockRepo.calls, ['getWarehouseInventory']);
      expect(mockRepo.lastInventoryParams, isNull);
    });

    test('propagates error from repository', () {
      mockRepo.errorToThrow = Exception('fetch error');

      expect(
        () => useCases.getWarehouseInventory(1, null),
        throwsA(isA<Exception>()),
      );
    });
  });

  group('adjustStock', () {
    test('delegates to repository and returns result', () async {
      mockRepo.adjustStockResult = _makeMovement(id: 99);

      final dto = AdjustStockDTO(
        productId: 'prod-1',
        warehouseId: 5,
        quantity: 10,
        reason: 'recount',
      );
      final result = await useCases.adjustStock(dto, businessId: 3);

      expect(mockRepo.calls, ['adjustStock']);
      expect(mockRepo.lastAdjustDTO, dto);
      expect(mockRepo.lastBusinessId, 3);
      expect(result.id, 99);
    });

    test('propagates error from repository', () {
      mockRepo.errorToThrow = Exception('adjust error');

      expect(
        () => useCases.adjustStock(
          AdjustStockDTO(
            productId: 'p1',
            warehouseId: 1,
            quantity: 1,
            reason: 'r',
          ),
        ),
        throwsA(isA<Exception>()),
      );
    });
  });

  group('transferStock', () {
    test('delegates to repository and returns result', () async {
      mockRepo.transferStockResult = {'success': true};

      final dto = TransferStockDTO(
        productId: 'prod-1',
        fromWarehouseId: 1,
        toWarehouseId: 2,
        quantity: 50,
      );
      final result = await useCases.transferStock(dto, businessId: 7);

      expect(mockRepo.calls, ['transferStock']);
      expect(mockRepo.lastTransferDTO, dto);
      expect(mockRepo.lastBusinessId, 7);
      expect(result['success'], true);
    });

    test('propagates error from repository', () {
      mockRepo.errorToThrow = Exception('transfer error');

      expect(
        () => useCases.transferStock(
          TransferStockDTO(
            productId: 'p1',
            fromWarehouseId: 1,
            toWarehouseId: 2,
            quantity: 1,
          ),
        ),
        throwsA(isA<Exception>()),
      );
    });
  });

  group('getMovements', () {
    test('delegates to repository and returns result', () async {
      mockRepo.getMovementsResult = PaginatedResponse<StockMovement>(
        data: [_makeMovement()],
        pagination: _makePagination(),
      );

      final params = GetMovementsParams(page: 1, pageSize: 10);
      final result = await useCases.getMovements(params);

      expect(mockRepo.calls, ['getMovements']);
      expect(mockRepo.lastMovementsParams, params);
      expect(result.data.length, 1);
    });

    test('delegates with null params', () async {
      mockRepo.getMovementsResult = PaginatedResponse<StockMovement>(
        data: [],
        pagination: _makePagination(),
      );

      await useCases.getMovements(null);

      expect(mockRepo.calls, ['getMovements']);
      expect(mockRepo.lastMovementsParams, isNull);
    });

    test('propagates error from repository', () {
      mockRepo.errorToThrow = Exception('movements error');

      expect(
        () => useCases.getMovements(null),
        throwsA(isA<Exception>()),
      );
    });
  });

  group('getMovementTypes', () {
    test('delegates to repository and returns result', () async {
      mockRepo.getMovementTypesResult = PaginatedResponse<MovementType>(
        data: [_makeMovementType()],
        pagination: _makePagination(),
      );

      final params = GetMovementTypesParams(activeOnly: true);
      final result = await useCases.getMovementTypes(params);

      expect(mockRepo.calls, ['getMovementTypes']);
      expect(mockRepo.lastMovementTypesParams, params);
      expect(result.data.length, 1);
    });

    test('delegates with null params', () async {
      mockRepo.getMovementTypesResult = PaginatedResponse<MovementType>(
        data: [],
        pagination: _makePagination(),
      );

      await useCases.getMovementTypes(null);

      expect(mockRepo.calls, ['getMovementTypes']);
      expect(mockRepo.lastMovementTypesParams, isNull);
    });

    test('propagates error from repository', () {
      mockRepo.errorToThrow = Exception('types error');

      expect(
        () => useCases.getMovementTypes(null),
        throwsA(isA<Exception>()),
      );
    });
  });
}
